File: Toggly.FeatureManagement.Python/toggly-django/src/toggly_django/middleware.py
"""Django middleware for Toggly feature flags."""

from __future__ import annotations

from typing import Callable

from django.http import HttpRequest, HttpResponse

from toggly_django.utils import get_client, get_context_from_request


class TogglyMiddleware:
    """Middleware that attaches Toggly context to each request.

    Adds `request.toggly` attribute for easy access to feature flags.

    Usage in settings.py:
        MIDDLEWARE = [
            ...
            'toggly_django.middleware.TogglyMiddleware',
            ...
        ]

    Usage in views:
        def my_view(request):
            if request.toggly.is_enabled('my-feature'):
                ...
    """

    def __init__(
        self,
        get_response: Callable[[HttpRequest], HttpResponse],
    ) -> None:
        """Initialize the middleware.

        Args:
            get_response: The next middleware or view.
        """
        self.get_response = get_response

    def __call__(self, request: HttpRequest) -> HttpResponse:
        """Process the request.

        Args:
            request: The incoming request.

        Returns:
            The response from the view.
        """
        # Attach Toggly helper to request
        request.toggly = TogglyRequestHelper(request)  # type: ignore

        response = self.get_response(request)

        return response


class TogglyRequestHelper:
    """Helper class attached to request for easy feature flag access."""

    def __init__(self, request: HttpRequest) -> None:
        """Initialize the helper.

        Args:
            request: The Django request.
        """
        self._request = request
        self._context = None
        self._client = get_client()

    @property
    def context(self):
        """Get or create the evaluation context for this request."""
        if self._context is None:
            self._context = get_context_from_request(self._request)
        return self._context

    def is_enabled(self, feature_key: str, default: bool = False) -> bool:
        """Check if a feature is enabled.

        Args:
            feature_key: The feature key to check.
            default: Default value if client unavailable.

        Returns:
            True if the feature is enabled.
        """
        if self._client is None:
            return default
        return self._client.is_enabled(feature_key, self.context, default=default)

    def is_disabled(self, feature_key: str, default: bool = True) -> bool:
        """Check if a feature is disabled.

        Args:
            feature_key: The feature key to check.
            default: Default value if client unavailable.

        Returns:
            True if the feature is disabled.
        """
        return not self.is_enabled(feature_key, default=not default)

    def evaluate_gate(
        self,
        feature_keys: list[str],
        requirement: str = "all",
        negate: bool = False,
    ) -> bool:
        """Evaluate multiple features as a gate.

        Args:
            feature_keys: List of feature keys.
            requirement: 'all' or 'any'.
            negate: Whether to negate the result.

        Returns:
            True if the gate passes.
        """
        if self._client is None:
            return False

        from toggly import FeatureRequirement

        req = (
            FeatureRequirement.ALL
            if requirement.lower() == "all"
            else FeatureRequirement.ANY
        )
        return self._client.evaluate_gate(
            feature_keys, req, self.context, negate
        )

    @property
    def flags(self) -> dict[str, bool]:
        """Get all feature flags.

        Returns:
            Dictionary of feature key to enabled state.
        """
        if self._client is None:
            return {}
        return self._client.feature_flags

File: Toggly.FeatureManagement.Node/toggly-express/src/middleware.ts
import type { Request, Response, NextFunction, RequestHandler } from 'express'
import {
  createTogglyClient,
  normalizeFeatureKeys,
  fromHttpRequest,
  type TogglyClient,
  type EvaluationContext,
} from '@ops-ai/toggly-node-core'
import type {
  TogglyExpressConfig,
  TogglyRequest,
  FeatureGateOptions,
  FeatureRouteOptions,
} from './types.js'

// Module-level client singleton
let expressClient: TogglyClient | null = null

/**
 * Get the Express Toggly client
 */
export function getExpressToggly(): TogglyClient | null {
  return expressClient
}

/**
 * Get identity from request
 */
async function extractIdentity(
  req: Request,
  config: TogglyExpressConfig
): Promise<string | undefined> {
  // Use custom extractor if provided
  if (config.getIdentity) {
    return config.getIdentity(req)
  }

  // Default: check header, then session
  const headerIdentity = req.headers['x-toggly-identity']
  if (typeof headerIdentity === 'string') {
    return headerIdentity
  }

  // Check session if available
  const session = (req as Request & { session?: { userId?: string } }).session
  if (session?.userId) {
    return session.userId
  }

  return undefined
}

/**
 * Get evaluation context from request
 */
async function extractContext(
  req: Request,
  config: TogglyExpressConfig
): Promise<EvaluationContext> {
  const headers = req.headers as Record<string, string | string[] | undefined>
  const headerRequest = fromHttpRequest(headers).request

  // Custom full context: use returned fields, fill missing request from headers
  if (config.getContext) {
    const custom = await config.getContext(req)
    return {
      ...custom,
      request: custom.request ?? headerRequest,
    }
  }

  // Default: identity / groups / claims providers + segment request headers
  const identity = await extractIdentity(req, config)
  const session = (req as Request & { session?: { groups?: string[] } }).session
  const groups = config.getGroups ? await config.getGroups(req) : session?.groups
  const claims = config.getClaims ? await config.getClaims(req) : undefined
  const fromReq = fromHttpRequest(headers, {
    identity,
    groups,
    claims,
  })

  return {
    identity: fromReq.identity,
    groups: fromReq.groups,
    claims: fromReq.claims,
    request: fromReq.request,
    traits: {
      ip: req.ip,
      userAgent: req.headers['user-agent'],
      path: req.path,
      method: req.method,
    },
  }
}

/**
 * Create Toggly middleware for Express
 *
 * This middleware:
 * 1. Initializes the Toggly client (if not already initialized)
 * 2. Attaches feature flag helpers to the request object
 * 3. Extracts identity and context from the request
 */
export function togglyMiddleware(config: TogglyExpressConfig): RequestHandler {
  // Initialize client lazily
  let initPromise: Promise<void> | null = null

  const initialize = async () => {
    if (!expressClient) {
      const {
        onError: _onError,
        getIdentity: _getIdentity,
        getGroups: _getGroups,
        getClaims: _getClaims,
        getContext: _getContext,
        ...serverConfig
      } = config
      expressClient = createTogglyClient(serverConfig)
      await expressClient.init()
    }
  }

  return async (req: Request, res: Response, next: NextFunction) => {
    try {
      // Ensure client is initialized
      if (!initPromise) {
        initPromise = initialize()
      }
      await initPromise

      if (!expressClient) {
        throw new Error('Toggly client failed to initialize')
      }

      // Extract context from request
      const context = await extractContext(req, config)
      const identity = context.identity

      // Attach Toggly helpers to request
      const togglyReq = req as TogglyRequest
      togglyReq.toggly = {
        client: expressClient,
        features: expressClient.state.features,
        identity,
        context,
        isFeatureOn: (featureKey: string) =>
          expressClient!.isFeatureOn(featureKey, context),
        isFeatureOff: (featureKey: string) =>
          expressClient!.isFeatureOff(featureKey, context),
        evaluateFeatureGate: (featureKeys, requirement, negate) =>
          expressClient!.evaluateFeatureGate(featureKeys, requirement, negate, context),
      }

      next()
    } catch (error) {
      if (config.onError) {
        config.onError(error as Error, req, res, next)
      } else {
        next(error)
      }
    }
  }
}

/**
 * Create a feature gate middleware
 *
 * Protects routes by checking if specified features are enabled
 */
export function featureGate(options: FeatureGateOptions): RequestHandler {
  const {
    featureKey,
    requirement = 'all',
    negate = false,
    onDisabled,
    redirectTo,
    redirectStatus = 302,
  } = options

  const featureKeys = normalizeFeatureKeys(featureKey)

  return async (req: Request, res: Response, next: NextFunction) => {
    const togglyReq = req as TogglyRequest

    if (!togglyReq.toggly) {
      // Toggly middleware not applied
      return next(new Error('Toggly middleware must be applied before featureGate'))
    }

    try {
      const isEnabled = await togglyReq.toggly.evaluateFeatureGate(
        featureKeys,
        requirement,
        negate
      )

      if (isEnabled) {
        return next()
      }

      // Feature is disabled
      if (redirectTo) {
        return res.redirect(redirectStatus, redirectTo)
      }

      if (onDisabled) {
        return onDisabled(req, res, next)
      }

      // Default: 404
      return res.status(404).json({
        error: 'Not Found',
        message: 'The requested resource is not available',
      })
    } catch (error) {
      return next(error)
    }
  }
}

/**
 * Create middleware that applies feature gates based on route patterns
 */
export function featureRoutes(routes: FeatureRouteOptions[]): RequestHandler {
  return async (req: Request, res: Response, next: NextFunction) => {
    for (const route of routes) {
      // Check if route matches
      const pathMatches =
        typeof route.path === 'string'
          ? req.path === route.path || req.path.startsWith(route.path)
          : route.path.test(req.path)

      if (!pathMatches) {
        continue
      }

      // Check if method matches
      if (route.methods && !route.methods.includes(req.method.toUpperCase())) {
        continue
      }

      // Apply feature gate
      const gate = featureGate(route)
      return gate(req, res, next)
    }

    // No matching route, continue
    return next()
  }
}

/**
 * Create a route handler that only executes if feature is enabled
 */
export function withFeature(
  featureKey: string | string[],
  handler: RequestHandler,
  options: Omit<FeatureGateOptions, 'featureKey'> = {}
): RequestHandler {
  const gate = featureGate({ featureKey, ...options })

  return async (req: Request, res: Response, next: NextFunction) => {
    return gate(req, res, (err?: unknown) => {
      if (err) {
        return next(err)
      }
      return handler(req, res, next)
    })
  }
}

/**
 * Get features as JSON endpoint handler
 */
export function featuresHandler(): RequestHandler {
  return (req: Request, res: Response) => {
    const togglyReq = req as TogglyRequest

    if (!togglyReq.toggly) {
      return res.status(500).json({
        error: 'Internal Server Error',
        message: 'Toggly middleware not configured',
      })
    }

    return res.json({
      features: togglyReq.toggly.features,
      identity: togglyReq.toggly.identity,
    })
  }
}

/**
 * Close the Express Toggly client
 */
export function closeExpressToggly(): void {
  if (expressClient) {
    expressClient.close()
    expressClient = null
  }
}
